diff -Naur a/drivers/staging/rtl8723bs/hal/rtl8723bs_recv.c b/drivers/staging/rtl8723bs/hal/rtl8723bs_recv.c
--- a/drivers/staging/rtl8723bs/hal/rtl8723bs_recv.c	2019-08-04 14:54:04.332245000 -0600
+++ b/drivers/staging/rtl8723bs/hal/rtl8723bs_recv.c	2019-08-04 14:55:10.559803837 -0600
@@ -109,7 +109,7 @@
 	rx_bssid = get_hdr_bssid(wlanhdr);
 	pkt_info.bssid_match = ((!IsFrameTypeCtrl(wlanhdr)) &&
 				!pattrib->icv_err && !pattrib->crc_err &&
-				!ether_addr_equal(rx_bssid, my_bssid));
+				ether_addr_equal(rx_bssid, my_bssid));
 
 	rx_ra = get_ra(wlanhdr);
 	my_hwaddr = myid(&padapter->eeprompriv);
